### ISSUE TITLE
T2.AlterantRebuilder.LateUpdate: IndexOutOfRangeException: index + length > size

### ISSUE THREAD

**Issue by ChrisNZL** [ChrisNZL/Tallowmere2#741]
A couple auto reports from v0.2.0 and v0.2.1.

_Looks_ like an internal Unity or Mono error when calling `List.Clear()` when alterants are being rebuilt from `T2.AlterantRebuilder.LateUpdate`.

---

Feedback ID: 20201210-NQB49

```
7:31:38, Frame 207884, EXCEPTION »  IndexOutOfRangeException: index + length > size
>>>>> CRITICAL ERROR >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

System.Array.Clear (System.Array array, System.Int32 index, System.Int32 length)
System.Collections.Generic.List`1[T].Clear ()
T2.AlterantTracker.RebuildAlterants ()
T2.AlterantTracker.RecalculateAlterantsViaAlterantRebuilder (System.Boolean rebuild, T2.AlterantRebuildFilter filter)
T2.AlterantRebuilder.LateUpdate ()

GameStates: RompingThroughDungeon
GameSetupMode: Singleplayer
Players: 1
DungeonRoomTransitionerState: Null
Room: 2 / 2
RoomModifiers: None
YouHaveDiedState: Null
SystemPlayer InputDevice: 键盘
HumanPlayer 1 InputDevice: 键盘
```

---

Feedback ID: 20201214-RR47T

```
8:06:52, Frame 402771, EXCEPTION »  IndexOutOfRangeException: index + length > size
>>>>> CRITICAL ERROR >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

System.Array.Clear (System.Array array, System.Int32 index, System.Int32 length)
System.Collections.Generic.List`1[T].Clear ()
T2.CoreAttribute_HealthOnKill.HardRecalculate (System.Collections.Generic.List`1[T] trackerAlterants, T2.IAlterantUser alterantUser)
T2.AlterantTracker.RecalculateAlterantsViaAlterantRebuilder (System.Boolean rebuild, T2.AlterantRebuildFilter filter)
T2.AlterantRebuilder.LateUpdate ()

GameStates: RompingThroughDungeon
GameSetupMode: Singleplayer
Players: 1
DungeonRoomTransitionerState: Null
Room: 10 / 10
RoomModifiers: None
YouHaveDiedState: Null
SystemPlayer InputDevice: 键盘
HumanPlayer 1 InputDevice: 键盘
```

Feedback ID: 20201216-FMK6P

```
BeginTransitioningToRoom: Departing room 5...


9:49:53, Frame 38010, EXCEPTION »  IndexOutOfRangeException: index + length > size
>>>>> CRITICAL ERROR >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

System.Array.Clear (System.Array array, System.Int32 index, System.Int32 length)
System.Collections.Generic.List`1[T].Clear ()
T2.CoreAttribute_ChestFind.HardRecalculate (System.Collections.Generic.List`1[T] trackerAlterants, T2.IAlterantUser alterantUser)
T2.AlterantTracker.RecalculateAlterantsViaAlterantRebuilder (System.Boolean rebuild, T2.AlterantRebuildFilter filter)
T2.AlterantRebuilder.LateUpdate ()

GameStates: RompingThroughDungeon, TransitioningToDungeonRoom
GameSetupMode: Singleplayer
Players: 1
DungeonRoomTransitionerState: WaitingForRoomToLoad
Room: 6 / 6
RoomModifiers: ManyElites (超多精英怪)
YouHaveDiedState: Null
SystemPlayer InputDevice: 键盘
HumanPlayer 1 InputDevice: 键盘
```

**Comment by ChrisNZL** [comment]
Still happening in 0.2.2.

Disabling Unity's "Physics 2D > Use Multithreading" option in 0.2.2 has not resolved this.

May need to look at try/catches over multiple frames, and/or ensure Lists aren't getting cleared or manipulated within a `foreach` somewhere.

Feedback ID: 20201225-3QLGJ

```
5:48:25, Frame 136679, EXCEPTION »  IndexOutOfRangeException: index + length > size
>>>>> CRITICAL ERROR >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

System.Array.Clear (System.Array array, System.Int32 index, System.Int32 length)
System.Collections.Generic.List`1[T].Clear ()
T2.AlterantTracker.RebuildAlterants ()
T2.AlterantTracker.RecalculateAlterantsViaAlterantRebuilder (System.Boolean rebuild, T2.AlterantRebuildFilter filter)
T2.AlterantRebuilder.LateUpdate ()

GameStates: RompingThroughDungeon, UsingMerchantMenu
OnlineGameInfo: Device is offline.
No previous server nor online game info.
GameSetupMode: Singleplayer
Players: 1
DungeonRoomTransitionerState: Null
Room: 0 / 1
RoomModifiers: None
YouHaveDiedState: Null
SystemPlayer InputDevice: 键盘
HumanPlayer 1 InputDevice: 键盘
```

**Comment by ChrisNZL** [comment]
Should be fixed in 0.2.3.

Added a try/catch within `AlterantTracker.RebuildAlterants` when calling `alterants.Clear`. If it fails due to a Unity error, a new List is now created.
